File: programs/dfs.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <unordered_set>

class Graph {
private:
    std::unordered_map<int, std::vector<int>> adj_list;

    void dfs_util(int u, std::unordered_set<int>& visited, std::vector<int>& result) {
        visited.insert(u);
        result.push_back(u);

        for (int v : adj_list[u]) {
            if (visited.count(v) == 0) {
                dfs_util(v, visited, result);
            }
        }
    }

public:
    void add_edge(int u, int v) {
        adj_list[u].push_back(v);
        adj_list[v].push_back(u);
    }

    std::vector<int> dfs(int start) {
        std::unordered_set<int> visited;
        std::vector<int> result;
        dfs_util(start, visited, result);
        return result;
    }
};

int main() {
    Graph graph;
    graph.add_edge(1, 2);
    graph.add_edge(1, 3);
    graph.add_edge(2, 4);
    graph.add_edge(2, 5);
    graph.add_edge(3, 6);

    std::vector<int> dfs_result = graph.dfs(1);
    for (int node : dfs_result) {
        std::cout << node << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: programs/dfs.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::hash::Hash;

struct Graph<T: Eq + Hash + Copy> {
    adj_list: HashMap<T, Vec<T>>,
}

impl<T: Eq + Hash + Copy> Graph<T> {
    fn new() -> Graph<T> {
        Graph {
            adj_list: HashMap::new(),
        }
    }

    fn add_edge(&mut self, u: T, v: T) {
        self.adj_list.entry(u).or_default().push(v);
        self.adj_list.entry(v).or_default().push(u);
    }

    fn dfs(&self, start: T) -> Vec<T> {
        let mut visited = HashSet::new();
        let mut result = Vec::new();
        self.dfs_util(start, &mut visited, &mut result);
        result
    }

    fn dfs_util(&self, u: T, visited: &mut HashSet<T>, result: &mut Vec<T>) {
        visited.insert(u);
        result.push(u);

        if let Some(neighbors) = self.adj_list.get(&u) {
            for &v in neighbors {
                if !visited.contains(&v) {
                    self.dfs_util(v, visited, result);
                }
            }
        }
    }
}

fn main() {
    let mut graph = Graph::new();
    graph.add_edge(1, 2);
    graph.add_edge(1, 3);
    graph.add_edge(2, 4);
    graph.add_edge(2, 5);
    graph.add_edge(3, 6);

    let dfs_result = graph.dfs(1);
    println!("DFS: {:?}", dfs_result);
}